File: archive/pulumi-py/Pr6160/lib/compute_stack.py
"""
compute_stack.py

Compute infrastructure with Auto Scaling Group, launch template, and IAM roles.
Instance types are environment-specific: t3.micro (dev), t3.small (staging), t3.medium (prod).
"""

from typing import List

import pulumi
from pulumi import Output, ResourceOptions
import pulumi_aws as aws


class ComputeStack(pulumi.ComponentResource):
    """
    Compute infrastructure component.

    Creates Auto Scaling Group with environment-appropriate instance types,
    launch template, IAM roles, and security groups.
    """

    def __init__(
        self,
        name: str,
        environment_suffix: str,
        vpc_id: Output[str],
        private_subnet_ids: List[Output[str]],
        alb_security_group_id: Output[str],
        alb_target_group_arn: Output[str],
        tags: dict,
        opts: ResourceOptions = None
    ):
        super().__init__('tap:compute:ComputeStack', name, None, opts)

        # Environment-specific instance types
        instance_types = {
            'dev': 't3.micro',
            'staging': 't3.small',
            'prod': 't3.medium'
        }
        instance_type = instance_types.get(environment_suffix, 't3.micro')

        # Environment-specific scaling configuration
        scaling_config = {
            'dev': {'min': 1, 'max': 2, 'desired': 1},
            'staging': {'min': 2, 'max': 4, 'desired': 2},
            'prod': {'min': 2, 'max': 6, 'desired': 2}
        }
        scaling = scaling_config.get(environment_suffix, scaling_config['dev'])

        # Create security group for instances
        self.instance_sg = aws.ec2.SecurityGroup(
            f'instance-sg-{environment_suffix}',
            vpc_id=vpc_id,
            description=f'Security group for EC2 instances in {environment_suffix}',
            ingress=[
                aws.ec2.SecurityGroupIngressArgs(
                    description='Allow HTTP from ALB',
                    from_port=80,
                    to_port=80,
                    protocol='tcp',
                    security_groups=[alb_security_group_id],
                ),
                aws.ec2.SecurityGroupIngressArgs(
                    description='Allow HTTPS from ALB',
                    from_port=443,
                    to_port=443,
                    protocol='tcp',
                    security_groups=[alb_security_group_id],
                ),
            ],
            egress=[
                aws.ec2.SecurityGroupEgressArgs(
                    description='Allow all outbound traffic',
                    from_port=0,
                    to_port=0,
                    protocol='-1',
                    cidr_blocks=['0.0.0.0/0'],
                ),
            ],
            tags={**tags, 'Name': f'instance-sg-{environment_suffix}'},
            opts=ResourceOptions(parent=self)
        )

        # Create IAM role for EC2 instances
        self.instance_role = aws.iam.Role(
            f'instance-role-{environment_suffix}',
            assume_role_policy="""{
                "Version": "2012-10-17",
                "Statement": [{
                    "Effect": "Allow",
                    "Principal": {"Service": "ec2.amazonaws.com"},
                    "Action": "sts:AssumeRole"
                }]
            }""",
            tags={**tags, 'Name': f'instance-role-{environment_suffix}'},
            opts=ResourceOptions(parent=self)
        )

        # Attach SSM policy for management
        aws.iam.RolePolicyAttachment(
            f'instance-role-ssm-{environment_suffix}',
            role=self.instance_role.name,
            policy_arn='arn:aws:iam::aws:policy/AmazonSSMManagedInstanceCore',
            opts=ResourceOptions(parent=self)
        )

        # Attach CloudWatch policy
        aws.iam.RolePolicyAttachment(
            f'instance-role-cloudwatch-{environment_suffix}',
            role=self.instance_role.name,
            policy_arn='arn:aws:iam::aws:policy/CloudWatchAgentServerPolicy',
            opts=ResourceOptions(parent=self)
        )

        # Create instance profile
        self.instance_profile = aws.iam.InstanceProfile(
            f'instance-profile-{environment_suffix}',
            role=self.instance_role.name,
            tags={**tags, 'Name': f'instance-profile-{environment_suffix}'},
            opts=ResourceOptions(parent=self)
        )

        # Get latest Amazon Linux 2023 AMI
        ami = aws.ec2.get_ami(
            most_recent=True,
            owners=['amazon'],
            filters=[
                aws.ec2.GetAmiFilterArgs(
                    name='name',
                    values=['al2023-ami-*-x86_64']
                ),
                aws.ec2.GetAmiFilterArgs(
                    name='virtualization-type',
                    values=['hvm']
                ),
            ]
        )

        # User data script
        user_data = (
            f"#!/bin/bash\n"
            f"set -e\n"
            f"yum update -y\n"
            f"yum install -y httpd\n"
            f"systemctl start httpd\n"
            f"systemctl enable httpd\n"
            f'echo "<h1>Environment: {environment_suffix}</h1>" > '
            f'/var/www/html/index.html\n'
        )

        # Create launch template
        self.launch_template = aws.ec2.LaunchTemplate(
            f'launch-template-{environment_suffix}',
            name_prefix=f'lt-{environment_suffix}-',
            image_id=ami.id,
            instance_type=instance_type,
            iam_instance_profile=aws.ec2.LaunchTemplateIamInstanceProfileArgs(
                arn=self.instance_profile.arn
            ),
            vpc_security_group_ids=[self.instance_sg.id],
            user_data=pulumi.Output.secret(user_data).apply(
                lambda ud: __import__('base64').b64encode(ud.encode()).decode()
            ),
            tag_specifications=[
                aws.ec2.LaunchTemplateTagSpecificationArgs(
                    resource_type='instance',
                    tags={**tags, 'Name': f'instance-{environment_suffix}'}
                ),
                aws.ec2.LaunchTemplateTagSpecificationArgs(
                    resource_type='volume',
                    tags={**tags, 'Name': f'volume-{environment_suffix}'}
                ),
            ],
            tags={**tags, 'Name': f'launch-template-{environment_suffix}'},
            opts=ResourceOptions(parent=self)
        )

        # Create Auto Scaling Group
        self.asg = aws.autoscaling.Group(
            f'asg-{environment_suffix}',
            name=f'asg-{environment_suffix}',
            min_size=scaling['min'],
            max_size=scaling['max'],
            desired_capacity=scaling['desired'],
            health_check_type='ELB',
            health_check_grace_period=300,
            vpc_zone_identifiers=private_subnet_ids,
            target_group_arns=[alb_target_group_arn],
            launch_template=aws.autoscaling.GroupLaunchTemplateArgs(
                id=self.launch_template.id,
                version='$Latest'
            ),
            tags=[
                aws.autoscaling.GroupTagArgs(
                    key=k,
                    value=v,
                    propagate_at_launch=True
                ) for k, v in {**tags, 'Name': f'asg-{environment_suffix}'}.items()
            ],
            opts=ResourceOptions(parent=self)
        )

        # Create scaling policy - target tracking based on CPU
        # pylint: disable=line-too-long
        predefined_metric = aws.autoscaling.PolicyTargetTrackingConfigurationPredefinedMetricSpecificationArgs(
            predefined_metric_type='ASGAverageCPUUtilization'
        )
        aws.autoscaling.Policy(
            f'asg-policy-cpu-{environment_suffix}',
            autoscaling_group_name=self.asg.name,
            policy_type='TargetTrackingScaling',
            target_tracking_configuration=aws.autoscaling.PolicyTargetTrackingConfigurationArgs(
                predefined_metric_specification=predefined_metric,
                target_value=70.0
            ),
            opts=ResourceOptions(parent=self)
        )

        # Export properties
        self.instance_security_group_id = self.instance_sg.id
        self.asg_name = self.asg.name
        self.asg_arn = self.asg.arn

        self.register_outputs({
            'instance_security_group_id': self.instance_security_group_id,
            'asg_name': self.asg_name,
            'asg_arn': self.asg_arn,
        })

File: archive/pulumi-ts/Pr6004/lib/compute-stack.ts
/**
 * compute-stack.ts
 *
 * Application Load Balancer and Auto Scaling Group infrastructure.
 */
import * as pulumi from '@pulumi/pulumi';
import * as aws from '@pulumi/aws';

export interface ComputeStackArgs {
  environmentSuffix: string;
  vpc: aws.ec2.Vpc;
  publicSubnetIds: pulumi.Output<string>[];
  privateSubnetIds: pulumi.Output<string>[];
  databaseEndpoint: pulumi.Output<string>;
  tags: pulumi.Input<{ [key: string]: string }>;
}

export class ComputeStack extends pulumi.ComponentResource {
  public readonly albArn: pulumi.Output<string>;
  public readonly albDnsName: pulumi.Output<string>;
  public readonly targetGroupArn: pulumi.Output<string>;
  public readonly asgName: pulumi.Output<string>;
  public readonly instanceIds: pulumi.Output<string[]>;

  constructor(
    name: string,
    args: ComputeStackArgs,
    opts?: pulumi.ComponentResourceOptions
  ) {
    super('tap:compute:ComputeStack', name, args, opts);

    const { environmentSuffix, vpc, publicSubnetIds, privateSubnetIds, tags } =
      args;

    // Security Group for ALB
    const albSecurityGroup = new aws.ec2.SecurityGroup(
      `alb-sg-${environmentSuffix}`,
      {
        vpcId: vpc.id,
        description: 'Security group for Application Load Balancer',
        ingress: [
          {
            protocol: 'tcp',
            fromPort: 80,
            toPort: 80,
            cidrBlocks: ['0.0.0.0/0'],
            description: 'Allow HTTP from anywhere',
          },
          {
            protocol: 'tcp',
            fromPort: 443,
            toPort: 443,
            cidrBlocks: ['0.0.0.0/0'],
            description: 'Allow HTTPS from anywhere',
          },
        ],
        egress: [
          {
            protocol: '-1',
            fromPort: 0,
            toPort: 0,
            cidrBlocks: ['0.0.0.0/0'],
            description: 'Allow all outbound traffic',
          },
        ],
        tags: {
          Name: `alb-sg-${environmentSuffix}`,
          ...tags,
        },
      },
      { parent: this }
    );

    // Security Group for EC2 instances
    const instanceSecurityGroup = new aws.ec2.SecurityGroup(
      `instance-sg-${environmentSuffix}`,
      {
        vpcId: vpc.id,
        description: 'Security group for EC2 instances',
        ingress: [
          {
            protocol: 'tcp',
            fromPort: 80,
            toPort: 80,
            securityGroups: [albSecurityGroup.id],
            description: 'Allow HTTP from ALB',
          },
        ],
        egress: [
          {
            protocol: '-1',
            fromPort: 0,
            toPort: 0,
            cidrBlocks: ['0.0.0.0/0'],
            description: 'Allow all outbound traffic',
          },
        ],
        tags: {
          Name: `instance-sg-${environmentSuffix}`,
          ...tags,
        },
      },
      { parent: this }
    );

    // Application Load Balancer
    const alb = new aws.lb.LoadBalancer(
      `alb-${environmentSuffix}`,
      {
        loadBalancerType: 'application',
        subnets: publicSubnetIds,
        securityGroups: [albSecurityGroup.id],
        enableCrossZoneLoadBalancing: true,
        enableHttp2: true,
        tags: {
          Name: `alb-${environmentSuffix}`,
          ...tags,
        },
      },
      { parent: this }
    );

    // CloudWatch Log Group for ALB
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const _albLogGroup = new aws.cloudwatch.LogGroup(
      `alb-logs-${environmentSuffix}`,
      {
        retentionInDays: 30,
        tags: {
          Name: `alb-logs-${environmentSuffix}`,
          ...tags,
        },
      },
      { parent: this }
    );

    // Target Group
    const targetGroup = new aws.lb.TargetGroup(
      `tg-${environmentSuffix}`,
      {
        port: 80,
        protocol: 'HTTP',
        vpcId: vpc.id,
        targetType: 'instance',
        deregistrationDelay: 30,
        healthCheck: {
          enabled: true,
          path: '/health',
          protocol: 'HTTP',
          port: '80',
          interval: 15,
          timeout: 5,
          healthyThreshold: 2,
          unhealthyThreshold: 2,
          matcher: '200',
        },
        tags: {
          Name: `tg-${environmentSuffix}`,
          ...tags,
        },
      },
      { parent: this }
    );

    // ALB Listener
    new aws.lb.Listener(
      `alb-listener-${environmentSuffix}`,
      {
        loadBalancerArn: alb.arn,
        port: 80,
        protocol: 'HTTP',
        defaultActions: [
          {
            type: 'forward',
            targetGroupArn: targetGroup.arn,
          },
        ],
      },
      { parent: this }
    );

    // Get latest Amazon Linux 2023 AMI
    const ami = aws.ec2.getAmiOutput({
      mostRecent: true,
      owners: ['amazon'],
      filters: [
        {
          name: 'name',
          values: ['al2023-ami-*-x86_64'],
        },
        {
          name: 'virtualization-type',
          values: ['hvm'],
        },
      ],
    });

    // IAM Role for EC2 instances
    const instanceRole = new aws.iam.Role(
      `instance-role-${environmentSuffix}`,
      {
        assumeRolePolicy: JSON.stringify({
          Version: '2012-10-17',
          Statement: [
            {
              Effect: 'Allow',
              Principal: { Service: 'ec2.amazonaws.com' },
              Action: 'sts:AssumeRole',
            },
          ],
        }),
        managedPolicyArns: [
          'arn:aws:iam::aws:policy/AmazonSSMManagedInstanceCore',
          'arn:aws:iam::aws:policy/CloudWatchAgentServerPolicy',
        ],
        tags: {
          Name: `instance-role-${environmentSuffix}`,
          ...tags,
        },
      },
      { parent: this }
    );

    const instanceProfile = new aws.iam.InstanceProfile(
      `instance-profile-${environmentSuffix}`,
      {
        role: instanceRole.name,
        tags: {
          Name: `instance-profile-${environmentSuffix}`,
          ...tags,
        },
      },
      { parent: this }
    );

    // CloudWatch Log Group for EC2
    const ec2LogGroup = new aws.cloudwatch.LogGroup(
      `ec2-logs-${environmentSuffix}`,
      {
        retentionInDays: 30,
        tags: {
          Name: `ec2-logs-${environmentSuffix}`,
          ...tags,
        },
      },
      { parent: this }
    );

    // User data script
    const userData = pulumi.interpolate`#!/bin/bash
set -e

# Update packages
yum update -y

# Install CloudWatch agent
yum install -y amazon-cloudwatch-agent

# Configure CloudWatch agent for application logs
cat > /opt/aws/amazon-cloudwatch-agent/etc/amazon-cloudwatch-agent.json <<'CFGEOF'
{
  "logs": {
    "logs_collected": {
      "files": {
        "collect_list": [
          {
            "file_path": "/var/log/app.log",
            "log_group_name": "${ec2LogGroup.name}",
            "log_stream_name": "{instance_id}/app.log",
            "timestamp_format": "%Y-%m-%d %H:%M:%S"
          }
        ]
      }
    }
  }
}
CFGEOF

# Start CloudWatch agent
/opt/aws/amazon-cloudwatch-agent/bin/amazon-cloudwatch-agent-ctl \
  -a fetch-config \
  -m ec2 \
  -s \
  -c file:/opt/aws/amazon-cloudwatch-agent/etc/amazon-cloudwatch-agent.json

# Install application dependencies
yum install -y httpd

# Create simple health check endpoint
echo "OK" > /var/www/html/health

# Start application
systemctl enable httpd
systemctl start httpd

# Create sample application log
touch /var/log/app.log
echo "$(date): Application started on $(hostname)" >> /var/log/app.log
`;

    // Launch Template
    const launchTemplate = new aws.ec2.LaunchTemplate(
      `launch-template-${environmentSuffix}`,
      {
        imageId: ami.id,
        instanceType: 't3.medium',
        iamInstanceProfile: {
          arn: instanceProfile.arn,
        },
        vpcSecurityGroupIds: [instanceSecurityGroup.id],
        userData: userData.apply(ud => Buffer.from(ud).toString('base64')),
        monitoring: {
          enabled: true,
        },
        metadataOptions: {
          httpEndpoint: 'enabled',
          httpTokens: 'required',
          httpPutResponseHopLimit: 1,
        },
        tagSpecifications: [
          {
            resourceType: 'instance',
            tags: {
              Name: `instance-${environmentSuffix}`,
              ...tags,
            },
          },
        ],
      },
      { parent: this }
    );

    // Auto Scaling Group
    const asg = new aws.autoscaling.Group(
      `asg-${environmentSuffix}`,
      {
        vpcZoneIdentifiers: privateSubnetIds,
        minSize: 3,
        maxSize: 9,
        desiredCapacity: 3,
        healthCheckType: 'ELB',
        healthCheckGracePeriod: 300,
        targetGroupArns: [targetGroup.arn],
        launchTemplate: {
          id: launchTemplate.id,
          version: '$Latest',
        },
        tags: [
          {
            key: 'Name',
            value: `asg-${environmentSuffix}`,
            propagateAtLaunch: true,
          },
        ],
      },
      { parent: this }
    );

    // Auto Scaling Policy - Scale Up
    const scaleUpPolicy = new aws.autoscaling.Policy(
      `scale-up-policy-${environmentSuffix}`,
      {
        autoscalingGroupName: asg.name,
        adjustmentType: 'ChangeInCapacity',
        scalingAdjustment: 1,
        cooldown: 300,
        policyType: 'SimpleScaling',
      },
      { parent: this }
    );

    // Auto Scaling Policy - Scale Down
    const scaleDownPolicy = new aws.autoscaling.Policy(
      `scale-down-policy-${environmentSuffix}`,
      {
        autoscalingGroupName: asg.name,
        adjustmentType: 'ChangeInCapacity',
        scalingAdjustment: -1,
        cooldown: 300,
        policyType: 'SimpleScaling',
      },
      { parent: this }
    );

    // CloudWatch Alarm - High CPU
    new aws.cloudwatch.MetricAlarm(
      `high-cpu-alarm-${environmentSuffix}`,
      {
        name: `high-cpu-alarm-${environmentSuffix}`,
        comparisonOperator: 'GreaterThanThreshold',
        evaluationPeriods: 2,
        metricName: 'CPUUtilization',
        namespace: 'AWS/EC2',
        period: 60,
        statistic: 'Average',
        threshold: 70,
        dimensions: {
          AutoScalingGroupName: asg.name,
        },
        alarmActions: [scaleUpPolicy.arn],
        tags: {
          Name: `high-cpu-alarm-${environmentSuffix}`,
          ...tags,
        },
      },
      { parent: this }
    );

    // CloudWatch Alarm - Low CPU
    new aws.cloudwatch.MetricAlarm(
      `low-cpu-alarm-${environmentSuffix}`,
      {
        name: `low-cpu-alarm-${environmentSuffix}`,
        comparisonOperator: 'LessThanThreshold',
        evaluationPeriods: 2,
        metricName: 'CPUUtilization',
        namespace: 'AWS/EC2',
        period: 60,
        statistic: 'Average',
        threshold: 30,
        dimensions: {
          AutoScalingGroupName: asg.name,
        },
        alarmActions: [scaleDownPolicy.arn],
        tags: {
          Name: `low-cpu-alarm-${environmentSuffix}`,
          ...tags,
        },
      },
      { parent: this }
    );

    // Lifecycle Hook for connection draining
    new aws.autoscaling.LifecycleHook(
      `termination-hook-${environmentSuffix}`,
      {
        autoscalingGroupName: asg.name,
        lifecycleTransition: 'autoscaling:EC2_INSTANCE_TERMINATING',
        defaultResult: 'CONTINUE',
        heartbeatTimeout: 300,
        name: `termination-hook-${environmentSuffix}`,
      },
      { parent: this }
    );

    // Export values
    this.albArn = alb.arn;
    this.albDnsName = alb.dnsName;
    this.targetGroupArn = targetGroup.arn;
    this.asgName = asg.name;
    this.instanceIds = pulumi.output([]);

    this.registerOutputs({
      albArn: this.albArn,
      albDnsName: this.albDnsName,
      targetGroupArn: this.targetGroupArn,
      asgName: this.asgName,
    });
  }
}
